In [47]:
#importing libararies 

import numpy as np 
import pandas as pd 
from sklearn.preprocessing import StandardScaler #used for standardization of the data 
from sklearn.model_selection import train_test_split #used for training and testing the model 
from sklearn import svm #used for support vector machine algo . 
from sklearn.metrics import accuracy_score 
#checking the accuracy of the model

DATA COLLECTION AND ANALYSIS 
PIMA DIABETES DATASET 

In [28]:
#loading the diabetes dataset to a pandas dataframe 
diabetes_db=pd.read_csv("/Users/sakshichaudhary/Downloads/diabetes.csv")

diabetes_db.head(5) #reading first 5 columns of the data 

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [29]:
diabetes_db.tail(10) #reading last 10 columns of the data using tail 

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
758            1      106             76              0        0  37.5   
759            6      190             92              0        0  35.5   
760            2       88             58             26       16  28.4   
761            9      170             74             31        0  44.0   
762            9       89             62              0        0  22.5   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
758                     0.197   26        0  
759                     0.278   66        1  
760                     0.766   22        0  
761                     0.403   43        1  
762                     0.142   33        0  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

In [30]:
#shows no. rows and columns present in the dataset 
diabetes_db.shape

(768, 9)

Diabetes_db.describe() is a function in Python that provides a statistical summary of a pandas DataFrame.

count: the number of non-null values in each column
mean: the average value of each column
std: the standard deviation of each column
min: the minimum value in each column
25%: the value at the 25th percentile of each column
50%: the value at the 50th percentile (also known as the median) of each column
75%: the value at the 75th percentile of each column
max: the maximum value in each column

For example, if diabetes_db is a DataFrame with columns for glucose levels, blood pressure, and age of patients with diabetes, diabetes_db.describe() will provide information about the distribution of glucose levels, blood pressure, and age in the dataset.

Note that describe() will only provide a summary of the numerical columns in the DataFrame. To include all columns, you can use :- 

diabetes_db.describe(include='all')

In [31]:
diabetes_db.dtypes #shows the datatypes of each column

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [32]:
#geting the statistical measures of the data using describe().
diabetes_db.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [33]:
diabetes_db["Outcome"].value_counts()
 
#500 people are non diabetic (0)
#268 people are diabetic (1) 

0    500
1    268
Name: Outcome, dtype: int64

In [34]:
 diabetes_db.describe(include='all')

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [35]:
 diabetes_db.groupby("Outcome").mean()
    
 #the diabetic people have more glucose in their blood 
#in this project GLUCOSE will be the main parameter to categorize :
#diabetic and non-diabetic people 
#i.e., outcome=1 are diabetic people  

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

#SEPARATING DATA  AND LABELS :-

- LABELS ARE THE OUTPUT predicted by various types of data in the ML model.
- Data are the values given to the ML model to train,test and validiation data ,Data are the input features of the model.

#When the data and labels are separated :-
1.On the basis of columns AXIS =1 ;
2.On the basis of rows AXIS=0;

#"drop" function is used to separate the data from particular row/column.


In [36]:
X=diabetes_db.drop("Outcome",axis=1)
print(X) #outcome column is dropped from table 


     Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  


In [37]:
#data is getting added in a new VARIABLE Y 
#to call a function we use square brackets [] 
#and not parenthesis() since it is not callable .

Y=diabetes_db["Outcome"] #a new column for outcome created .
print(Y)

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64


DATA STANDARDIZATION 
- In SVM it is important to standardize the data to get accurate results.
- It helps in putting in best category.

In [38]:
#creating an object for scaling  
scaler=StandardScaler()

In [39]:
#fit() is used to fit the data for further scaling.
scaler.fit(X)

StandardScaler()

In [40]:
standardized_data=scaler.transform(X)
print(standardized_data)


[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


In [41]:
X=standardized_data #represents the data which is standardized 
print(X)
print()
Y=diabetes_db["Outcome"]
print(Y) #y represents the Model 

#all the data is in X and all the labels are stored in Y 

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64




# AFTER TRANSFORMING THE DATA 
- we need to split the data into our training data on text Data 
- so TRAIN TEST SPLIT is used .

--> This is an example of how to use the train_test_split function in Python to split a dataset into training and testing sets. Here is an explanation of the different parameters used in this example:

- X: This is the input data that will be split into training and testing sets.
- Y: This is the target variable that will also be split into training and testing sets.

- Test_size: This parameter specifies the proportion of the dataset that will be used for testing. In this example, 20% of the dataset will be used for testing.

- Stratify: This parameter ensures that the training and testing sets have a proportional distribution of the target variable. In this case, Y is used as the stratification variable to ensure that the classes in Y are represented proportionally in both the training and testing sets.

- Random_state: This parameter is used to control the randomization applied to the dataset during the split. In this example, a random state of 2 is used to ensure that the same split will be obtained each time the code is run.

# The output of this function will be 4 arrays:

- X_train: This is the input data for the training set.
- X_test: This is the input data for the testing set.
- Y_train: This is the target variable for the training set.
- Y_test: This is the target variable for the testing set.

--> These arrays can then be used to train a machine learning model on the training data and evaluate its performance on the testing data.

In [42]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,
                                                 stratify=Y, random_state=2)
 

In [43]:
print(X.shape)
print(X_train.shape)
print(X_test.shape) 

#here the shape of data is defined for :-
#1.original data X.shape
#2.training data X_train.shape
#3.testing data  X_test,shape

(768, 8)
(614, 8)
(154, 8)



- Out of 768 examples in our original data ,out of those 614 is used for training the data and out of which 20% as the test size is taken=0.2,of the data i.e., 154 is used for testing of the data 


# TRAINING THE MODEL :



#creating a variable classifier 
#SVC represents support vector classifier 
#svm is library used for SUPPPORT VECTOR MACHINE ALGORITHM

# A new parameter is created known as "kernel"

- It creates a model that can classify data into different categories based on their features, using a linear decision boundary.

- The 'kernel' parameter specifies the type of kernel function used to transform the data into a higher-dimensional space where a linear boundary can be used to classify the data.

- Here, the 'Linear' kernel indicates that no such transformation is performed and the classification is done in the original feature space.

- A linear kernel is suitable when the data is already linearly separable, meaning the classes can be separated by a straight line.

- classifier will try to find the best hyperplane that separates the data points of different classes in the original feature space(2D SPACE) without transforming the data into a higher-dimensional space.


In [48]:
 classifier =  svm.SVC(kernel='linear') 
    #linear is always written in lowercase

In [49]:
classifier.fit(X_train,Y_train) 

SVC(kernel='linear')

# MODEL EVALUATION 
# ACCURACY SCORE ON TRAINING DATA 

- X_train_prediction=classifier.predict(X_train) generates predicted labels for the training data X_train using the trained SVM classifier classifier. The predicted labels are stored in the variable X_train_prediction.

- training_data_accuracy = accuracy_score(X_train_prediction,Y_train) calculates the accuracy of the trained SVM classifier on the training data.

- accuracy_score() function from the scikit-learn library is used to compare the predicted labels X_train_prediction to the true labels Y_train and calculate the accuracy. 
- The resulting accuracy value is 78% which is stored in the variable training_data_accuracy.


In [51]:
X_train_prediction=classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction,Y_train) 

In [52]:
print("Accuracy score of the training Data : ",training_data_accuracy)  

#we have acheived 78% accuracy 

Accuracy score of the training Data :  0.7866449511400652


# ACCURACY ON THE TESTING DATA-

- The code is predicting the output values for a set of "testing data (X_test)" using a machine learning model (classifier). 
- Then, it compares the predicted output values (X_test_prediction) with the actual output values (Y_test) for the same set of test data using a metric called accuracy score.
- Accuracy score is a metric that measures how often the model's prediction is correct. 
- 77 % is a good accuracy which shows that there is no overfitting done in the model ,hence the model is performing well on the testing data as well .


In [55]:
X_test_prediction=classifier.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction,Y_test)

In [57]:
print("Accuracy score of testing data : ",testing_data_accuracy)

Accuracy score of testing data :  0.7727272727272727


# MAKING A PREDICTIVE SYSTEM FOR DIABETES 



In [87]:
input_data=(8,99,84,0,0,35.4,0.388,50)

#label is not included as we are predicting the output 
#for input_data in input_data_list:

#change the input data to NUMPY ARRAY .
input_data_as_numpy_array = np.asarray(input_data)

#reshape the array as we are predicting for only one instance 
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

#standardize the input data 
std_data = scaler.transform(input_data_reshaped)
print(std_data)
print()

prediction=classifier.predict(std_data)
print("RESULT: " ,prediction)

if(prediction[0] == 0):
    print("The Person is NON-DIABETIC")
else:
    print("The Person is DIABETIC")

[[ 1.23388019 -0.68523633  0.77001375 -1.28821221 -0.69289057  0.43246741
  -0.25331639  1.4259954 ]]

RESULT:  [0]
The Person is NON-DIABETIC


/Users/sakshichaudhary/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(
